File: src/tool/cesium/meteorology/typhoon/typhoon.ts
import { Cartesian3, Color, EllipseGeometry, EllipsoidSurfaceAppearance, GeometryInstance, Material, PolygonGeometry, PolygonHierarchy, Primitive, VertexFormat, Math as CesiumMath } from "cesium";
// @ts-ignore
import typhoonFs from './typhoon.fs?raw';
import { Position, RectangleRange } from "../../type/RectangleRange";

type TyphoonRadius = {
    NE12: number,
    NW12: number,
    SE12: number,
    SW12: number,
    NE10: number,
    NW10: number,
    SE10: number,
    SW10: number,
    NE7: number,
    NW7: number,
    SE7: number,
    SW7: number,
}

class TyphoonMaterial extends Material {
    typhoonRadiusData: TyphoonRadius;
    constructor(typhoonRadius: TyphoonRadius, colorString: string = '#0aa1ed') {
        let arr = [typhoonRadius.NE12, typhoonRadius.NW12, typhoonRadius.SE12, typhoonRadius.SW12, typhoonRadius.NE10, typhoonRadius.NW10, typhoonRadius.SE10, typhoonRadius.SW10, typhoonRadius.NE7, typhoonRadius.NW7, typhoonRadius.SE7, typhoonRadius.SW7];
        let max = Math.max(...arr);
        if (max == 0) {
            throw new Error("台风数据错误：最大半径值为0");
        }
        arr.map((item, index) => {
            arr[index] = item / max;
        });
        super({
            fabric: {
                type: 'TyphoonMaterial',
                uniforms: {
                    color: Color.fromCssColorString(colorString),
                    NE12: arr[0],
                    NW12: arr[1],
                    SE12: arr[2],
                    SW12: arr[3],
                    NE10: arr[4],
                    NW10: arr[5],
                    SE10: arr[6],
                    SW10: arr[7],
                    NE7: arr[8],
                    NW7: arr[9],
                    SE7: arr[10],
                    SW7: arr[11],
                },
                source: typhoonFs
            },
            translucent: true,
        });
        this.typhoonRadiusData = typhoonRadius;
    }
    setColor(color: Color) {
        this.uniforms.color = color;
    }

    setRadius(typhoonRadius: TyphoonRadius) {
        let arr = [typhoonRadius.NE12, typhoonRadius.NW12, typhoonRadius.SE12, typhoonRadius.SW12, typhoonRadius.NE10, typhoonRadius.NW10, typhoonRadius.SE10, typhoonRadius.SW10, typhoonRadius.NE7, typhoonRadius.NW7, typhoonRadius.SE7, typhoonRadius.SW7];
        let max = Math.max(...arr);
        if (max == 0) {
            throw new Error("台风数据错误：最大半径值为0");
        }
        arr.map((item, index) => {
            arr[index] = item / max;
        });
        // 以下是十二个方向的半径赋值
        this.uniforms.NE12 = arr[0];
        this.uniforms.NW12 = arr[1];
        this.uniforms.SE12 = arr[2];
        this.uniforms.SW12 = arr[3];
        this.uniforms.NE10 = arr[4];
        this.uniforms.NW10 = arr[5];
        this.uniforms.SE10 = arr[6];
        this.uniforms.SW10 = arr[7];
        this.uniforms.NE7 = arr[8];
        this.uniforms.NW7 = arr[9];
        this.uniforms.SE7 = arr[10];
        this.uniforms.SW7 = arr[11];
    }
}


export class TyphoonPrimitive extends Primitive {
    constructor(position: Position, typhoonRadius: TyphoonRadius, colorString: string = '#0aa1ed') {
        const typhoonMaterial = new TyphoonMaterial(typhoonRadius, colorString);
        let arr = [typhoonRadius.NE12, typhoonRadius.NW12, typhoonRadius.SE12, typhoonRadius.SW12, typhoonRadius.NE10, typhoonRadius.NW10, typhoonRadius.SE10, typhoonRadius.SW10, typhoonRadius.NE7, typhoonRadius.NW7, typhoonRadius.SE7, typhoonRadius.SW7];
        let max = Math.max(...arr);
        super({
            geometryInstances: new GeometryInstance({
                geometry: new EllipseGeometry({
                    center: Cartesian3.fromDegrees(position.lon, position.lat),
                    semiMinorAxis: max,
                    semiMajorAxis: max,
                    rotation: CesiumMath.PI_OVER_FOUR,
                    vertexFormat: VertexFormat.POSITION_AND_ST
                }),
            }),
            appearance: new EllipsoidSurfaceAppearance({
                aboveGround: false,
                material: typhoonMaterial,
            }),
        });
    }
    
}

File: src/tool/cesium/meteorology/typhoon/typhoon.fs
uniform vec4 color;
uniform float NW12;
uniform float NE12;
uniform float SE12;
uniform float SW12;
uniform float NW10;
uniform float NE10;
uniform float SE10;
uniform float SW10;
uniform float NW7;
uniform float NE7;
uniform float SE7;
uniform float SW7;

float dwind(vec2 uv, vec4 ri, vec3 co) {
    float re = 0.;
    if(uv.x * co.x > 0. && uv.y * co.y > 0. && ri.y <= ri.x) {
        if(co.z > 0.) {
            re = ri.w;
        }
        if(ri.y > ri.x - 2. * ri.b) {
            re = 1.;
        }
    }
    return re;
}

// 2维线段sdf函数其中参数：p：当前点位置，a：起点位置，b：终点位置,返回值：点渲染的颜色的透明度
float lineSdf(vec2 p, vec2 a, vec2 b) {
    vec2 ap = p - a;
    vec2 ab = b - a;
    float vl = dot(ap, ab) / dot(ab, ab);
    if(vl >= 0. && vl <= 1.) {
        vec2 hv = ap - vl * ab;
        float d = length(hv);
        if(d < fwidth(p.x)) {
            return 1.;
        } else {
            return 0.;
        }
    } else {
        return 0.;
    }
}

czm_material czm_getMaterial(czm_materialInput materialInput) {
    czm_material material = czm_getDefaultMaterial(materialInput);
    material.diffuse = color.rgb;
    vec2 uv = (materialInput.st - 0.5) * 2.;
    float d = length(uv);
    float w = fwidth(max(uv.x, uv.y));
    float apl = 0.;
    float ia = clamp(d, 0., .4);
    apl += dwind(uv, vec4(SW12, d, w, ia), vec3(-1., -1., 1.));
    apl += dwind(uv, vec4(NE12, d, w, ia), vec3(1., 1., 1.));
    apl += dwind(uv, vec4(NW12, d, w, ia), vec3(-1., 1., 1.));
    apl += dwind(uv, vec4(SE12, d, w, ia), vec3(1, -1., 1.));

    apl += dwind(uv, vec4(SW10, d, w, ia), vec3(-1., -1., -1.));
    apl += dwind(uv, vec4(NE10, d, w, ia), vec3(1., 1., -1.));
    apl += dwind(uv, vec4(NW10, d, w, ia), vec3(-1., 1., -1.));
    apl += dwind(uv, vec4(SE10, d, w, ia), vec3(1, -1., -1.));

    apl += dwind(uv, vec4(SW7, d, w, ia), vec3(-1., -1., -1.));
    apl += dwind(uv, vec4(NE7, d, w, ia), vec3(1., 1., -1.));
    apl += dwind(uv, vec4(NW7, d, w, ia), vec3(-1., 1., -1.));
    apl += dwind(uv, vec4(SE7, d, w, ia), vec3(1, -1., -1.));

    apl += lineSdf(uv, vec2(-NW12, 0.), vec2(-SW12, 0.));
    apl += lineSdf(uv, vec2(-NW10, 0.), vec2(-SW10, 0.));
    apl += lineSdf(uv, vec2(-NW7, 0.), vec2(-SW7, 0.));

    apl += lineSdf(uv, vec2(NE12, 0.), vec2(SE12, 0.));
    apl += lineSdf(uv, vec2(NE10, 0.), vec2(SE10, 0.));
    apl += lineSdf(uv, vec2(NE7, 0.), vec2(SE7, 0.));

    apl += lineSdf(uv, vec2(0., NW12), vec2(0., NE12));
    apl += lineSdf(uv, vec2(0., NW10), vec2(0., NE10));
    apl += lineSdf(uv, vec2(0., NW7), vec2(0., NE7));

    apl += lineSdf(uv, vec2(0., -SW12), vec2(0., -SE12));
    apl += lineSdf(uv, vec2(0., -SW10), vec2(0., -SE10));
    apl += lineSdf(uv, vec2(0., -SW7), vec2(0., -SE7));

    apl = clamp(apl, 0., 1.);
    material.alpha = apl;
    return material;
}